x/twap/listeners_test.go

```diff
@@ -329,6 +375,227 @@ func (s *TestSuite) TestPoolStateChange() {
+	// we assume that it is new. That is, we drop the state and knowledge of the
+	// last error time prior to that as well as any twap history.
+	twapAfterCreatePosition3 := validateUpdatedRecordDoesNotEqualPrevious(twapAfterWithdawPosition1, blockTimeSix)
+
+	////////////////////////////////////////////////////////////
+	// Eight block: swap after re-adding liquidity
+
+	s.RunBasicSwap(poolId)
+
+	s.validateChangedPool()
+
+	s.EndBlock()
+	s.Commit()
+
+	// Note that performing a swap changes twap
+	validateUpdatedRecordDoesNotEqualPrevious(twapAfterCreatePosition3, blockTimeSix)
+}
+
+// This test validates that all twap record mutators (listeners) run as expected
+// and update twap + last spot price error at the desired points in the execution flow.
+// It assumed that every state change message occurs within the same block.
+// This is in contrast to TestPoolStateChange_Concentrated_SeparateBlocks
+// that runs the same tests but within a separate block.
+func (s *TestSuite) TestPoolStateChange_Concentrated_SameBlock() {
+	s.SetupTest()
+
+	var (
+		////////////////////////////////////////////////////////////
+		// 1: create pool
+		clPool = s.PrepareConcentratedPoolWithCoins(denom0, denom1)
+		poolId = clPool.GetId()
+	)
+
+	firstBlockTime := s.Ctx.BlockTime()
+
+	////////////////////////////////////////////////////////////
+	// 2: create a first position that initializes twap record.
+
+	positionId, liquidityCreated := s.CreateFullRangePosition(clPool, defaultThreeAssetCoins)
+
+	////////////////////////////////////////////////////////////
+	// 3: create a second position that does not change twap record.
+
+	positionId2, liquidityCreated2 := s.CreateFullRangePosition(clPool, defaultThreeAssetCoins)
+
+	// Note that creating second position has no effect.
+
+	////////////////////////////////////////////////////////////
+	// 4: perform swap and update twap.
+
+	s.RunBasicSwap(poolId)
+
+	////////////////////////////////////////////////////////////
+	// 5: partial withdraw -> does not update twap.
+
+	s.WithdrawFullRangePosition(clPool, positionId2, liquidityCreated2)
+
+	////////////////////////////////////////////////////////////
+	// 6: withdraw in-full -> twap changes with error.
+
+	s.WithdrawFullRangePosition(clPool, positionId, liquidityCreated)
+
+	////////////////////////////////////////////////////////////
+	// 7: create new position, updating twap with valid spot price
+
+	s.CreateFullRangePosition(clPool, defaultThreeAssetCoins)
+
+	////////////////////////////////////////////////////////////
+	// 8: swap after re-adding liquidity
+
+	s.RunBasicSwap(poolId)
+
+	s.validateChangedPool()
+
+	s.EndBlock()
+	s.Commit()
+
+	// Note that when we create a concentrated liquidity pool in the same block with
+	// create initial position that correctly initializes spot price and twap,
+	// we still end up having the error time set to the block time when the pool was created.
+	s.validateRecordUpdated(firstBlockTime)
+}
```

Comment:
Would it be possible / do you think it makes sense to add test cases where we test the actual values of the twap from Concentrated Liquidity? something like this: https://github.com/osmosis-labs/osmosis/blob/32a285f4fc/x/twap/api_test.go#L179 All current tests are using Balancer pools to test correctness of the twap values.

I don't think it necessarily has to be included in the PR, please create an issue if you agree!

Reply:
The test you linked does not have knowledge of the pool. It operates on twap records. So there is no difference if records were created from a balancer or a CL pool. In the context of twap, there is no difference what pool it is tracking twap for **as long as the twap records are created/updated in the right places**  and **twap error is correctly propagated if it exists**.

The tests in `listeners_test.go` validate that.

So to answer your question, I don't think more tests are necessary.

I think an e2e test would be a nice-to-have but not a requirement for this PR.

Reply:
For e2e: https://github.com/osmosis-labs/osmosis/issues/4844